### PR TITLE
Update Linq, openAPI, bump versions of TS, Node

### PR DESCRIPTION
**Pull request by fearthecowboy**
Creating a branch to do a large update.

- LINQ - refactored some of this so that it's much better performance when it can be. A lot of the implementation can be done with arrays (and hence, added the linq-like names to those), but further, even iterating over things is faster if you don't force it to be linq all the time 

- OpenAPI - updated the OpenAPI types to include both OAI2 and OAI3. Much better version than last one

This will be breaking changes, so all versions are bumped, and consumers will have to bump their version to take these changes. 

- Moved to TS 3.9.5
- switched to testdeck/mocha from mocha-typescript
- Made low-end node version 12.x 
- bumped versions of everything
- updated some tests 

I'm going to set some CI up this branch to publish the packages pre-release so that consumers can try them out without us conflictng.

### CODE REVIEW

File: codegen/pluralization.ts [hunk 1]
```diff
@@ -1,5 +1,6 @@
 import { pascalCase, deconstruct } from './text-manipulation';
-
+import { linq } from '@azure-tools/linq';
+linq;
```
Comment (daviwil) [inline codegen/pluralization.ts:3]: Probably not needed?

```suggestion
```

File: linq/new-linq.ts [hunk 2]
```diff
@@ -111,75 +137,62 @@ function isIterable<T>(source: any): source is Iterable<T> {
...
-      }
-    }());
+    return Object.values(source);
   }
 
   // null/undefined
-  return linqify([]);
+  return [];
+}
+export const linq ={
```
Comment (daviwil) [inline linq/new-linq.ts:159]: Any particular reason for the namespacing?  Clarity/readability at call sites?

**Review by daviwil — APPROVED** [review]
This is great!  The improvements to `linq` will help a lot with perf and usability.  I also like the merging of OpenAPI 2 and 3 typings, they seem much more robust than the last iteration.

File: linq/new-linq.ts (hunk 2 above)
Comment (fearthecowboy) [reply linq/new-linq.ts:159]: To make using the `IterableWithLinq` functions a deliberate act -- the non namespaced versions are much more lightweigt.

File: codegen/pluralization.ts (hunk 1 above)
Comment (fearthecowboy) [reply codegen/pluralization.ts:3]: It needs a reference to linq for it to pick up the Array extensions.

**Pull request merged**